Use Privacy Pass, captcha, and Rack to rate limit sign ins and sign ups

mercedesb opened the pull request:
Resolves #3518 

## What was the end-user or developer problem that led to this PR?

Please see #3518 

Rubygems.org wants to reduce (and prevent if possible) bot activity around signup and login. Captchas are a common way to reduce activity, but can also be annoying and repetitive. We’d like to use Privacy Pass to get the same security with a smoother user experience then Captcha, although it doesn’t completely replace Captchas.

<!-- Write a clear and complete description of the problem -->

## What is your fix for the problem, implemented in this PR?

Following the product design in the linked issue, I've implemented privacy pass token challenges and redemption, hcaptcha verification, and stricter rate limiting around sign in and sign up.

## To test
Privacy pass is currently supported on iOS and MacOS Ventura. I tested this with Safari in MacOS Ventura 13.2.1 (it doesn't work in any non-Apple browser yet as far as I can tell).

You may need to create an SSL-terminating tunnel that reverse proxies to localhost:3000 (https://localhost.run/ or https://ngrok.com worked fine, just be sure to update the allowed hosts for dev).

## Deployment next steps
Added env vars (you may want to move some of these into Rails credentials 🤷‍♀️)
- `HCAPTCHA_SITE_KEY`: the site key for RubyGems' hcaptcha account
- `HCAPTCHA_SECRET`: the secret for RubyGems' hcaptcha account
- `PRIVACY_PASS_ISSUER_PUBLIC_KEY`: the Fastly privacy pass public key (currently falling back to their demo)
- `PRIVACY_PASS_ISSUER_NAME`: the Fastly privacy pass issuer host name (currently falling back to their demo)

## Other helpful info
[Apple has the best info currently](https://developer.apple.com/news/?id=huqjyh7k) about how to set up your web server to serve privacy pass challenges and redeem privacy pass tokens. 

The various docs around the IETF RFC are also helpful
- https://www.ietf.org/archive/id/draft-ietf-privacypass-auth-scheme-08.html#name-token-challenge
- https://www.ietf.org/archive/id/draft-ietf-privacypass-architecture-03.html#name-redemption-protocol
- https://www.ietf.org/archive/id/draft-ietf-privacypass-protocol-04.html#section-6.3

hcaptcha was straightforward to get set up and working. [Their docs](https://docs.hcaptcha.com/) are pretty good.

## Make sure the following tasks are checked

- [x] Describe the problem / feature
- [x] Write [tests](https://github.com/rubygems/rubygems/blob/master/bundler/doc/development/PULL_REQUESTS.md#tests) for features and bug fixes
- [x] Write code to solve the problem
- [x] Make sure you follow the [current code style](https://github.com/rubygems/rubygems/blob/master/bundler/doc/development/PULL_REQUESTS.md#code-formatting) and [write meaningful commit messages without tags](https://github.com/rubygems/rubygems/blob/master/bundler/doc/development/PULL_REQUESTS.md#commit-messages)

segiddins commented on config/routes.rb:
```diff
@@ -209,7 +210,11 @@
     end
 
-    get '/sign_in' => 'clearance/sessions#new', as: 'sign_in'
+    resource :user, only: [] do
```
is `resource only: []` any different from using `namespace` ?

mercedesb replied:
Good question, I was following the pattern that currently exists in `routes` for session and I think in this case it makes more sense. `/user` isn't a namespace but a singular resource. Namespacing it would mean that by convention we would expect the controller to exist within a user module.

From the Rails docs
> [2.5 Singular Resources](https://guides.rubyonrails.org/routing.html#singular-resources)
Sometimes, you have a resource that clients always look up without referencing an ID. ..
> [2.6 Controller Namespaces and Routing](https://guides.rubyonrails.org/routing.html#controller-namespaces-and-routing)
> You may wish to organize groups of controllers under a namespace. Most commonly, you might group a number of administrative controllers under an Admin:: namespace, and place these controllers under the app/controllers/admin directory. You can route to such a group by using a [namespace](https://api.rubyonrails.org/v7.0.4.2/classes/ActionDispatch/Routing/Mapper/Scoping.html#method-i-namespace) block:

segiddins commented on app/controllers/sessions_controller.rb:
```diff
@@ -1,20 +1,27 @@
+    if redeem_privacy_pass_token
+      super
+    else
+      setup_privacy_pass_challenge
```
nit: i think some of this might go better in a `before_action unless: :redeem_privacy_pass_token`

segiddins commented on app/controllers/sessions_controller.rb:
```diff
@@ -75,6 +82,19 @@ def mfa_create
+    if captcha_success
+      should_mfa ? webauthn_and_mfa_new : do_login
+    else
+      login_failure(t("captcha.invalid"))
```
nit: do this before checking `should_mfa`

segiddins commented on app/controllers/sessions_controller.rb:
```diff
@@ -75,6 +82,19 @@ def mfa_create
+    @user = User.find(session[:captcha_user])
+    session.delete(:captcha_user)
+
+    captcha_success = HcaptchaVerifier.call(captcha_response, request.remote_ip)
```
should we do this before we even find the user?

segiddins commented on app/controllers/users_controller.rb:
```diff
@@ -1,22 +1,63 @@
+    @user = user_from_params
+    session.delete(:captcha_user)
+
+    captcha_success = HcaptchaVerifier.call(captcha_response, request.remote_ip)
```
this seems repeated, maybe move into the concern?

segiddins commented on lib/hcaptcha_verifier.rb:
```diff
@@ -0,0 +1,68 @@
+      payload = {
+        response: client_response,
+        remoteip: remote_ip,
+        secret: ENV.fetch("HCAPTCHA_SECRET", "0x0000000000000000000000000000000000000000"),
```
these env var values should go into `config/deploy/*/secrets.ejson`, and then they should be added to the unicorn k8s container definition as env vars

mercedesb replied:
I don't have the actual values to use in deployed environments. I used only demo accounts for my development but this may require someone with creds for hcaptcha and Fastly to grab.

segiddins commented on lib/hcaptcha_verifier.rb:
```diff
@@ -0,0 +1,68 @@
+        :open_timeout   => 5,
+        "Content-Type"  => "application/x-www-form-urlencoded"
+
+      response_json = JSON.parse(response)
```
nit: request `application/json` via `Accepts` if we're assuming a JSON response

segiddins commented on lib/privacy_pass_redeemer.rb:
```diff
@@ -0,0 +1,111 @@
+  #   uint8_t token_key_id[32];
+  #   uint8_t authenticator[Nk];
+  # } Token;
+  TOKEN_TYPE_LENGTH = 1
```
should this be 2 since it's a `uint16_t` as opposed to a uint8?

mercedesb replied:
No, because we're checking the length unpacked data and we're unpacking this via the uint16 directive (`n`)

segiddins commented on lib/privacy_pass_redeemer.rb:
```diff
@@ -0,0 +1,111 @@
+
+  def challenge_digest_bytes
+    start_index = TOKEN_TYPE_LENGTH + NONCE_LENGTH
+    unpacked_token.slice(start_index, CHALLENGE_DIGEST_LENGTH)
```
instead of repeating this pattern a bunch, maybe set these all as ivars in the initializer? that way we can also do it all in order / keep the start index & lengths in sync?

segiddins commented on lib/privacy_pass_tokenizer.rb:
```diff
@@ -0,0 +1,89 @@
+  def self.issuer_public_key
+    # fastly demo fallback
+    # rubocop:disable Metrics/LineLength
+    ENV.fetch("PRIVACY_PASS_ISSUER_PUBLIC_KEY",
```
maybe link to where this value comes from?

indirect commented:
Fastly says we are in their beta access group now, and they are targeting end of Q2 for the beta issuer. In the meantime, I would be okay to ship this with just captchas now, and add privacy pass support once there's a production issuer. (Fastly says the demo issuer that we built this to work with is not for production use, so we shouldn't ship anything that depends on that.)

mercedesb commented:
> In the meantime, I would be okay to ship this with just captchas now, and add privacy pass support once there's a production issuer.

Based on @indirect's most recent feedback, I went ahead and feature flagged the privacy pass implementation using LaunchDarkly (someone with access to that will need to configure the variation)

segiddins commented on app/controllers/concerns/privacy_pass_supportable.rb:
```diff
@@ -0,0 +1,32 @@
+
+  def privacy_pass_enabled?
+    ld_context = LaunchDarkly::LDContext.with_key(self.class.name)
+    Rails.configuration.launch_darkly_client.variation("privacy_pass.enabled", ld_context, false)
```
nit: let's call the flag `"gemcutter.privacy_pass.enabled"`. should we use the current controller & action as the key, so we can test by only turning it on for a single action, rather than every action?

mercedesb replied:
I updated the name.

I think using the current controller would be good. My only concern about using the action is that for every place we want it (sign up or sign in) we actually need it enabled in 2 actions. The `new` action presents the challenge and the `create` action handles redeeming the token.

segiddins replied:
fair enough! I just think we'll need a `kind` on the LDContext of say `controller` so it doesn't look like a user

simi commented on app/views/sessions/captcha.html.erb:
```diff
@@ -0,0 +1,15 @@
+<div class="mfa__container">
+  <div class="mfa__option">
+    <%= form_tag captcha_create_session_path, method: :post, class: "mfa-form" do %>
+      <div class="h-captcha" data-sitekey="<%= ENV.fetch("HCAPTCHA_SITE_KEY", "10000000-ffff-ffff-ffff-000000000001") %>"></div>
```
Can you move `HCAPTCHA_SITE_KEY` to be ingested in `config/secrets.yml` from ENV (and specify this default for dev/testing environment if useful), introduce helper like `hcaptcha_site_key` to read secrets and use it in templates?

simi commented on lib/hcaptcha_verifier.rb:
```diff
@@ -0,0 +1,69 @@
+        sitekey: ENV.fetch("HCAPTCHA_SITE_KEY", "10000000-ffff-ffff-ffff-000000000001")
+      }
+
+      response = RestClient.post VERIFY_ENDPOINT,
```
RestClient was replaced by Faraday recently at https://github.com/rubygems/rubygems.org/pull/3602

simi commented on app/controllers/sessions_controller.rb:
```diff
@@ -172,6 +194,16 @@ def setup_mfa_authentication
   end
 
+  def setup_captcha_verification
+    create_catpcha_user(id: @user.id)
```
Per my local testing, captcha is also created when trying to login with unknown email, that means `@user` is nil in here and it crashes.

simi replied:
The same seems applied when using wrong password for existing account. `find_user` returns nil -> `@user` is nil.

simi commented:
I did some testing and left some initial comments. Regarding `ENV` values, please move them into `config/secrets.yml` (could be ingested from `ENV` still) or to `config/rubygems.yml` for fixed values.

Here's a screenshot from testing of captcha. The red text is just notice of using testing environment. It looks good to me.

![image](https://user-images.githubusercontent.com/193936/225851893-819f1f4f-7362-4e1c-896e-7238f62470d8.png)

mercedesb commented:
- put hcpatcha vals into secrets with testing creds as dev defaults
- replaced RestClient with Faraday
- fixed bug when user not found on login but captcha conditions met

simi commented on lib/privacy_pass_tokenizer.rb:
```diff
@@ -0,0 +1,90 @@
+    # fastly demo fallback
+    # https://www.fastly.com/blog/private-access-tokens-stepping-into-the-privacy-respecting-captcha-less#demo-configuration
+    # rubocop:disable Layout/LineLength
+    ENV.fetch("PRIVACY_PASS_ISSUER_PUBLIC_KEY",
```
Would it make sense to move this also to rubygems.yml/secrets.yml?

mercedesb replied:
I can. It's not really a secret but no harm in adding it there.

simi commented on app/controllers/concerns/privacy_pass_supportable.rb:
```diff
@@ -0,0 +1,32 @@
+
+  def valid_privacy_pass_redemption?
+    return false unless privacy_pass_enabled?
+    return session[:redeemed_privacy_pass] unless session[:redeemed_privacy_pass].nil?
```
I found out this a little hard to read during testing. Would it be possible to keep the condition "positive"?

```ruby
return session[:redeemed_privacy_pass] if session.key?(:redeemed_privacy_pass)
```

simi commented:
@mercedesb any more hints for local Privacy pass testing? I tried with [Privacy pass extension](https://privacypass.github.io/), but I couldn't get it to work :thinking:.

mercedesb commented:
> any more hints for local Privacy pass testing? I tried with [Privacy pass extension](https://privacypass.github.io/), but I couldn't get it to work

1. We're using the Fastly issuer and Fastly does not work with the Privacy Pass extension. That extension is a Cloudflare product.
2. I never had much luck testing with the extension anywhere on the web and neither did a colleague of mine. They have a lot of [open issues in their client repo](https://github.com/privacypass/challenge-bypass-extension/issues) with people having similar trouble and I don't have a lot of confidence in how well-supported the extension is.

Privacy Pass is currently supported natively only in iOS 16+, iPad 16+, macOS Ventura (13+). Through my testing, I had to use the Safari browser. But if you're running a supported OS with Safari, that should be all you need for testing.

If things are working as expected, you won't see the 401 response with the token challenge in the browser's dev tools' Network tab. You'll see a 200 with an Authorization header from the client which means the challenge was handled successfully. But you can step through the code to see the entire flow.

simi commented:
That extension supports also hCaptcha and seems active on hCaptcha rubygems.org page, but I couldn't get it to work as well. I don't have any Apple product around to test this.

mercedesb commented:
> That extension supports also hCaptcha and seems active on hCaptcha rubygems.org page

hcaptcha is a different issuer and won't work with Fastly tokens. We are using the Fastly issuer.

> I don't have any Apple product around to test this.

That's a bummer. Given that Apple seems to be the only native support for PATs at the moment, I do think it important that it's tested there.

simi commented on app/controllers/users_controller.rb:
```diff
@@ -1,22 +1,61 @@
+    elsif @user.save
+      handle_user_after_save
+    else
+      render template: "users/new"
```
If I understand it well, this is tricky to test. `valid?` is called few lines before, but this `save` can still fail. In theory codecov could be fixed by using `save!` and `rescue_from`, but I'm not sure if that's worth it. :thinking:

simi commented:
> That's a bummer. Given that Apple seems to be the only native support for PATs at the moment, I do think it important that it's tested there.

We can ask someone for help with testing on staging.

Other than that part I can't test, this looks good to me. :muscle:

simi commented on lib/hcaptcha_verifier.rb:
```diff
@@ -0,0 +1,75 @@
+      should_verify(keys, SIGN_UP_CAPTCHA_THRESHOLD)
+    end
+
+    def call(client_response, remote_ip)
```
_nitpick: I'm personally not big fan of `call` methods if not used as a proc. Would it make sense to rename this to something like `verified?`. Similar change could be done in PrivacyPassRedeemer._

simi commented on lib/hcaptcha_verifier.rb:
```diff
@@ -0,0 +1,75 @@
+        f.request :url_encoded
+        f.response :json
+        f.response :logger, logger, headers: false, errors: true
+        f.response :raise_error
```
Is this codecov complain related to stubs being used in tests instead of using test adapter? :thinking: 

https://lostisland.github.io/faraday/adapters/testing
